File: apps/backend/internal/models/distribution.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type DistributionStatus string

const (
	DistStatusScheduled DistributionStatus = "SCHEDULED"
	DistStatusPreparing DistributionStatus = "PREPARING"
	DistStatusInTransit DistributionStatus = "IN_TRANSIT"
	DistStatusDelivered DistributionStatus = "DELIVERED"
	DistStatusRejected  DistributionStatus = "REJECTED"
)

type BASTStatus string

const (
	BASTStatusGenerated BASTStatus = "GENERATED"
	BASTStatusSigned    BASTStatus = "SIGNED"
	BASTStatusArchived  BASTStatus = "ARCHIVED"
)

type DistributionPointType string

const (
	DPTypeSchool    DistributionPointType = "SCHOOL"    // PAUD, SD, SMP, SMA, SMK, SLB
	DPTypePosyandu  DistributionPointType = "POSYANDU"  // Posyandu untuk Kelompok 3B (Bumil, Busui, Balita)
	DPTypePesantren DistributionPointType = "PESANTREN" // Pondok Pesantren
)

type EducationLevel string

const (
	EduPAUD EducationLevel = "PAUD"
	EduSD   EducationLevel = "SD"
	EduSMP  EducationLevel = "SMP"
	EduSMA  EducationLevel = "SMA"
	EduSMK  EducationLevel = "SMK"
	EduSLB  EducationLevel = "SLB"
	EduNone EducationLevel = "NONE"
)

type PackageType string

const (
	PackageFoodTray PackageType = "FOOD_TRAY" // Siap Santap (Food Tray - hari aktif sekolah)
	PackageTotebag  PackageType = "TOTEBAG"   // Kemasan / Totebag (Libur sekolah / Kelompok 3B)
)

type DistributionPoint struct {
	AuditModel
	NPSN            string                `gorm:"type:varchar(32);uniqueIndex" json:"npsn,omitempty"`
	Name            string                `gorm:"type:varchar(255);not null" json:"name"`
	Type            DistributionPointType `gorm:"type:varchar(32);not null;default:'SCHOOL';index" json:"type"`
	EducationLevel  *EducationLevel       `gorm:"type:varchar(16)" json:"education_level,omitempty"`
	Address         string                `gorm:"type:text;not null" json:"address"`
	District        string                `gorm:"type:varchar(100);not null" json:"district"`
	City            string                `gorm:"type:varchar(100);not null" json:"city"`
	ContactPerson   string                `gorm:"type:varchar(150);not null" json:"contact_person"`
	PhoneNumber     string                `gorm:"type:varchar(50);not null" json:"phone_number"`
	TotalRecipients int                   `gorm:"type:integer;not null" json:"total_recipients"`
	DietaryNotes    string                `gorm:"type:text" json:"dietary_notes,omitempty"`
	Latitude        *float64              `gorm:"type:numeric(10,7)" json:"latitude,omitempty"`
	Longitude       *float64              `gorm:"type:numeric(10,7)" json:"longitude,omitempty"`
	IsActive        bool                  `gorm:"type:boolean;not null;default:true" json:"is_active"`
}

type Distribution struct {
	AuditModel
	DeliveryNumber      string             `gorm:"type:varchar(64);uniqueIndex;not null" json:"delivery_number"`
	DistributionPointID uuid.UUID          `gorm:"type:uuid;index" json:"distribution_point_id"`
	DistributionPoint   *DistributionPoint `gorm:"foreignKey:DistributionPointID" json:"distribution_point,omitempty"`
	DeliveryDate        time.Time          `gorm:"type:date;not null;index" json:"delivery_date"`
	Status              DistributionStatus `gorm:"type:varchar(32);not null;default:'SCHEDULED';index" json:"status"`
	PackageType         PackageType        `gorm:"type:varchar(32);not null;default:'FOOD_TRAY'" json:"package_type"`
	IsHolidayDelivery   bool               `gorm:"type:boolean;not null;default:false" json:"is_holiday_delivery"`
	DriverName          string             `gorm:"type:varchar(150);not null" json:"driver_name"`
	VehiclePlate        string             `gorm:"type:varchar(32);not null" json:"vehicle_plate"`
	TotalPortions       int                `gorm:"type:integer;not null" json:"total_portions"`
	TotalValue          float64            `gorm:"type:numeric(18,4);not null;default:0" json:"total_value"`
	RecipientName       string             `gorm:"type:varchar(150)" json:"recipient_name,omitempty"`
	RecipientTitle      string             `gorm:"type:varchar(100)" json:"recipient_title,omitempty"`
	ReceivedAt          *time.Time         `gorm:"type:timestamptz" json:"received_at,omitempty"`
	ProofOfDeliveryURL  string             `gorm:"type:text" json:"proof_of_delivery_url,omitempty"`
	Notes               string             `gorm:"type:text" json:"notes,omitempty"`
	Items               []DistributionItem `gorm:"foreignKey:DistributionID;constraint:OnDelete:CASCADE" json:"items,omitempty"`
}

type DistributionItem struct {
	AuditModel
	DistributionID   uuid.UUID  `gorm:"type:uuid;not null;index" json:"distribution_id"`
	MenuItemID       *uuid.UUID `gorm:"type:uuid" json:"menu_item_id,omitempty"`
	MealName         string     `gorm:"type:varchar(255);not null" json:"meal_name"`
	PortionsSent     int        `gorm:"type:integer;not null" json:"portions_sent"`
	PortionsReceived int        `gorm:"type:integer;not null;default:0" json:"portions_received"`
	UnitPrice        float64    `gorm:"type:numeric(15,4);not null;default:15000" json:"unit_price"`
	Subtotal         float64    `gorm:"type:numeric(18,4);not null" json:"subtotal"`
}

type BASTDocument struct {
	AuditModel
	DocumentNumber              string             `gorm:"type:varchar(64);uniqueIndex;not null" json:"document_number"`
	DistributionPointID         uuid.UUID          `gorm:"type:uuid;index" json:"distribution_point_id"`
	DistributionPoint           *DistributionPoint `gorm:"foreignKey:DistributionPointID" json:"distribution_point,omitempty"`
	PeriodStart                 time.Time          `gorm:"type:date;not null" json:"period_start"`
	PeriodEnd                   time.Time          `gorm:"type:date;not null" json:"period_end"`
	TotalPortions               int                `gorm:"type:integer;not null" json:"total_portions"`
	TotalAmount                 float64            `gorm:"type:numeric(18,4);not null" json:"total_amount"`
	FileURL                     string             `gorm:"type:text;not null" json:"file_url"`
	FileSizeBytes               int64              `gorm:"type:bigint" json:"file_size_bytes,omitempty"`
	GeneratedAt                 time.Time          `gorm:"type:timestamptz;not null" json:"generated_at"`
	SPPGHeadName                string             `gorm:"type:varchar(150);not null" json:"sppg_head_name"`
	RecipientRepresentativeName string             `gorm:"type:varchar(150);default:'Penerima Manfaat'" json:"recipient_representative_name"` // Kepsek / Bidan Desa / Pimpinan Ponpes
	Status                      BASTStatus         `gorm:"type:varchar(32);not null;default:'GENERATED'" json:"status"`
}

File: packages/shared-types/src/distribution.ts
import { AuditModel } from './audit';

export type DistributionPointType = 'SCHOOL' | 'POSYANDU' | 'PESANTREN';
export type EducationLevel = 'PAUD' | 'SD' | 'SMP' | 'SMA' | 'SMK' | 'SLB' | 'NONE';
export type PackageType = 'FOOD_TRAY' | 'TOTEBAG';

export interface DistributionPoint extends AuditModel {
  npsn?: string; // NPSN (School only)
  name: string;
  type: DistributionPointType;
  education_level?: EducationLevel;
  address: string;
  district: string;
  city: string;
  contact_person: string;
  phone_number: string;
  total_recipients: number;
  total_students?: number; // backward compat alias
  dietary_notes?: string;
  latitude?: number;
  longitude?: number;
  is_active: boolean;
}

// For backward compatibility
export type School = DistributionPoint;

export interface CreateDistributionPointRequest {
  npsn?: string;
  name: string;
  type: DistributionPointType;
  education_level?: EducationLevel;
  address: string;
  district: string;
  city: string;
  contact_person: string;
  phone_number: string;
  total_recipients: number;
  total_students?: number; // backward compat alias
  dietary_notes?: string;
  latitude?: number;
  longitude?: number;
}

export type CreateSchoolRequest = CreateDistributionPointRequest;

export type DistributionStatus = 'SCHEDULED' | 'PREPARING' | 'IN_TRANSIT' | 'DELIVERED' | 'REJECTED';

export interface CreateDistributionItemRequest {
  menu_item_id?: string;
  meal_name: string;
  portions_sent: number;
  unit_price?: number;
}

export interface CreateDistributionRequest {
  distribution_point_id: string;
  delivery_date: string;
  package_type?: PackageType;
  is_holiday_delivery?: boolean;
  driver_name: string;
  vehicle_plate: string;
  items: CreateDistributionItemRequest[];
  notes?: string;
}

export interface UpdateDistributionStatusRequest {
  status: DistributionStatus;
  recipient_name?: string;
  recipient_title?: string;
  proof_of_delivery_url?: string;
  items?: { item_id: string; portions_received: number }[];
  notes?: string;
}

export interface DistributionItem extends AuditModel {
  distribution_id: string;
  menu_item_id?: string;
  meal_name: string;
  portions_sent: number;
  portions_received: number;
  unit_price: number;
  subtotal: number;
}

export interface Distribution extends AuditModel {
  delivery_number: string;
  distribution_point_id: string;
  distribution_point?: DistributionPoint;
  school?: DistributionPoint; // backward compatibility
  delivery_date: string; // YYYY-MM-DD
  status: DistributionStatus;
  package_type: PackageType;
  is_holiday_delivery: boolean;
  driver_name: string;
  vehicle_plate: string;
  total_portions: number;
  total_value: number;
  recipient_name?: string;
  recipient_title?: string;
  received_at?: string;
  proof_of_delivery_url?: string;
  notes?: string;
  items: DistributionItem[];
}

export interface BASTDocument extends AuditModel {
  document_number: string;
  distribution_point_id: string;
  distribution_point?: DistributionPoint;
  school?: DistributionPoint; // backward compat
  period_start: string; // YYYY-MM-DD
  period_end: string; // YYYY-MM-DD
  total_portions: number;
  total_amount: number;
  file_url: string;
  file_size_bytes?: number;
  generated_at: string;
  sppg_head_name: string;
  recipient_representative_name: string;
  school_principal_name?: string; // backward compat
  status: 'GENERATED' | 'SIGNED' | 'ARCHIVED';
}

export interface BASTGenerateRequest {
  distribution_point_id?: string;
  school_id?: string; // backward compat
  period_start: string;
  period_end: string;
  sppg_head_name?: string;
  recipient_representative_name?: string;
  school_principal_name?: string; // backward compat
  official_notes?: string;
}

export interface BASTPreviewData {
  distribution_point: DistributionPoint;
  school?: DistributionPoint; // backward compat
  period_start: string;
  period_end: string;
  deliveries: Distribution[];
  total_deliveries: number;
  total_portions: number;
  total_amount: number;
}

export interface BASTDocumentResponse {
  id: string;
  document_number: string;
  distribution_point_id: string;
  distribution_point_name: string;
  school_id?: string; // backward compat
  school_name?: string; // backward compat
  period_start: string;
  period_end: string;
  total_portions: number;
  total_amount: number;
  file_url: string;
  file_size_bytes?: number;
  generated_at: string;
  sppg_head_name: string;
  recipient_representative_name: string;
  school_principal_name?: string; // backward compat
  status: 'GENERATED' | 'SIGNED' | 'ARCHIVED';
}
